akka-typed/src/main/scala/akka/typed/cluster/ClusterSharding.scala

```diff
@@ -0,0 +1,196 @@
+   */
+  def entityMessage(message: E): A
+
+  /**
+   * Extract the entity id from an incoming `message`. Only messages that passed the [[#entityId]]
+   * function will be used as input to this function.
+   */
+  def shardId(message: E): String
+}
+
+/**
+ * Java API:
+ *
+ * Default message extractor type, using envelopes to identify what entity a message is for
+ * and the hashcode of the entityId to decide which shard an entity belongs to.
+ *
+ * This is recommended since it does not force details about sharding into the entity protocol
+ *
+ * @tparam A The type of message accepted by the entity actor
+ */
+final class DefaultMessageExtractor[A](maxNumberOfShards: Int) extends TypedMessageExtractor[ShardingEnvelope[A], A] {
+  def entityId(envelope: ShardingEnvelope[A]) = envelope.entityId
+  def entityMessage(envelope: ShardingEnvelope[A]) = envelope.message
+  def shardId(envelope: ShardingEnvelope[A]) = (math.abs(envelope.entityId.hashCode) % maxNumberOfShards).toString
+}
+
+/**
+ * Java API:
+ *
+ * Default message extractor type, using a property of the message to identify what entity a message is for
+ * and the hashcode of the entityId to decide which shard an entity belongs to.
+ *
+ * This is recommended since it does not force details about sharding into the entity protocol
+ *
+ * @tparam A The type of message accepted by the entity actor
+ */
+abstract class DefaultNoEnvelopeMessageExtractor[A](maxNumberOfShards: Int) extends TypedMessageExtractor[A, A] {
+  def entityMessage(message: A) = message
+  def shardId(message: A) = {
+    val id = entityId(message)
+    if (id != null) (math.abs(id.hashCode) % maxNumberOfShards).toString
+    else null
+  }
+}
+
+/**
+ * A reference to an entityId and the local access to sharding, allows for actor-like interaction
+ *
+ * The entity ref must be resolved locally and cannot be sent to another node.
+ *
+ * TODO what about ask, should it actually implement ActorRef to be exactly like ActorRef and callers does not have
+ *      to know at all about it or is it good with a distinction but lookalike API?
+ */
+trait EntityRef[A] {
+  /**
+   * Send a message to the entity referenced by this EntityRef using *at-most-once*
+   * messaging semantics.
+   */
+  def tell(msg: A): Unit
+}
+
+object EntityRef {
+  implicit final class EntityRefOps[T](val ref: EntityRef[T]) extends AnyVal {
+    /**
+     * Send a message to the Actor referenced by this ActorRef using *at-most-once*
+     * messaging semantics.
+     */
+    def !(msg: T): Unit = ref.tell(msg)
+  }
+}
+
+object ClusterSharding extends ExtensionId[ClusterSharding] {
+  def createExtension(system: ActorSystem[_]): ClusterSharding = ???
+
+  /**
+   * Create an ActorRef-like reference to a specific sharded entity. Messages sent to it will be wrapped
+   * in a [[ShardingEnvelope]] and passed to the local shard region or proxy.
+   */
+  def entityRefFor[A](entityId: String, actorRef: ActorRef[ShardingEnvelope[A]]): EntityRef[A] =
```

Comment:
or move it to the extension instance, and have `typeName: String` param instead of `region: ActorRef` ?